### PR TITLE
Fix IPython kernel-related cleanup

### PR DESCRIPTION
**Pull request by mdickinson**
This PR shows a proof of concept for fixing cleanup issues in the InternalIPKernel and associated plugin. Should eventually address issues #94, #185, #186 and #187.

~Currently WIP, so making this a draft PR. Opening a PR now (a) to get CI feedback and (b) to have a place outside the code to record things that are still to do.~

Ready for review.

To do:

- [x] Lots of cleanup and removal of duplicated code.
- [x] Tests.
   - [x] Test that sys.std*, sys.excepthook and sys.displayhook are all restored as expected
   - [x] Test that IPython.utils.io globals unaltered
   - [x] Test that no ZMQInteractiveShell or IPythonKernel objects remain after cleanup
- [x] Absorb changes from #183 (they need extending to cover `rprint` and `rprinte`, which are still used in some versions of the code).
- [ ] zmq Sockets are only being cleaned up through cyclic gc. Can we clean them up better?
- [x] investigate and fix the `ResourceWarning` from `test_shutdown_closes_console_pipes`

Note: commit history is *very* messy. Please be sure to squash and merge.

### CODE REVIEW

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 1]
```diff
@@ -3,19 +3,52 @@
...
+#     for the rewrite.
+
+
+if six.PY2:
+    def atexit_unregister(func):
+        # Replace the contents, not the list itself, in case anyone else
+        # is keeping references to it.
+        atexit._exithandlers[:] = list(
+            handler for handler in atexit._exithandlers
+            if not handler[0] == func
```
Comment (jwiggins) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: Why this instead of `if handler[0] != func`? Something weird about this particular comparison?

Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: This is me being defensive: I'm deliberately trying to match the Python 3 C code as closely as possible, and that code uses an `__eq__` comparison rather than a `__ne__` comparison. It's not a given that the two are logical negations of each other: a class can override them both in incompatible ways, or in Python 2 (much more likely) override `__eq__` and neglect to provide the corresponding override for `__ne__`.

Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: > that code uses an `__eq__` comparison rather than a `__ne__` comparison

Code link: https://github.com/python/cpython/blob/2fc1160a80733f4c5c88796319154b3f59e98e4b/Modules/atexitmodule.c#L284

Comment (jwiggins) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: OK, I figured there was a reason. You should probably leave a note for future readers.

**Comment by mdickinson** [comment]
I'm getting a `ResourceWarning` from one of the tests:
```
test_shutdown_closes_console_pipes (envisage.plugins.ipython_kernel.tests.test_internal_ipkernel.TestInternalIPKernel) ... 
To connect another client to this kernel, use:
    --existing kernel-95701.json
/Users/mdickinson/.edm/envs/envisage-test-3.6-null/lib/python3.6/subprocess.py:766: ResourceWarning: subprocess 95708 is still running
  ResourceWarning, source=self)
ok
```

**Comment by mdickinson** [comment]
The `ResourceWarning` is a result of a semi-bug: when we clean up the consoles, we do a `console.term()` for each console, but don't then do a `console.wait()`; I've added that missing `wait` call, which sets the return code on the `Popen` object, which in turn allows the subprocess module to recognise that the child process has finished, and prevents the `ResourceWarning` from being issued.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 1 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Done.

**Comment by mdickinson** [comment]
This is ready for review. There's still a lingering puzzle about three `IOPubThread` sockets: I can't figure out how/where they're being closed, but they *are* being closed, so that's the main thing that matters. I'll pursue that outside this PR, though.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 2]
```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
...
+        self.assertTrue(all(socket.closed for socket in sockets))
+
+    def test_ipykernel_live_objects(self):
+        # Check that all IPython-related objects have been cleaned up
+        # as expected.
+
+        self.create_and_destroy_kernel()
+
+        # Sometimes once isn't enough ....
+        # XXX Ideally, this wouldn't be necessary at all; see if we can fix.
```
Comment (mdickinson) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Remove this comment. We're not going to go chasing all the cycles - there are way too many.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 3]
```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
...
+        kernels = self.objects_of_type(ipykernel.ipkernel.IPythonKernel)
+        self.assertEqual(kernels, [])
+
+        kernel_apps = self.objects_of_type(ipykernel.kernelapp.IPKernelApp)
+        self.assertEqual(kernel_apps, [])
+
+    def test_ctrl_c_message_suppressed(self):
+        captured = []
+        with capture_stream(sys.__stdout__, callback=captured.append):
+            kernel = InternalIPKernel()
```
Comment (mdickinson) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: could use `create_and_destroy_kernel` here.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 4]
```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
...
+        Find and return a list of all currently tracked instances of the
+        given type.
+        """
+        return [
+            obj for obj in gc.get_objects()
+            if isinstance(obj, type)
+        ]
+
+    def create_and_destroy_kernel(self):
+        kernel = InternalIPKernel()
```
Comment (mdickinson) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: This could use a docstring.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 5]
```diff
@@ -90,62 +117,51 @@ def init_ipkernel(self, gui_backend):
         # Start IPython kernel with GUI event loop support
         self.ipkernel = gui_kernel(gui_backend)
 
-        # Since ipykernel 4.7, the io_loop attribute of the kernel is not
-        # initialized anymore
-        # Reference: https://github.com/enthought/envisage/issues/107
-        # Workaround: Retrieve the kernel on the IPykernelApp and set the
-        # io_loop without starting it!
-        if NEEDS_IOLOOP_PATCH and not hasattr(self.ipkernel.kernel, 'io_loop'):
-            self.ipkernel.kernel.io_loop = ioloop.IOLoop.instance()
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: Can you comment on why this piece of code got removed? Was this fixed upstream?

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 6]
```diff
@@ -90,62 +117,51 @@ def init_ipkernel(self, gui_backend):
...
-                sys.stdout = self._original_stdout
-                kernel_stdout.close()
-
-            kernel_stderr = sys.stderr
-            if kernel_stderr is not self._original_stderr:
-                sys.stderr = self._original_stderr
-                kernel_stderr.close()
-
-            self.ipkernel.iopub_thread.stop()
+            self.ipkernel.shutdown()
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: Can you comment on how `ipkernel.shutdown` does the job of 

```
self.ipkernel.shell.exit_now = True
self.ipkernel.cleanup_connection_file()
self.ipkernel.iopub_thread.stop()
```

?

Comment (rahulporuri) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Or maybe the removed code isn't relevant anylonger or was the source of bugs.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 7]
```diff
@@ -15,59 +19,106 @@
...
+        self.records = []
+
+    def emit(self, record):
+        self.records.append(record)
+
+
 @contextlib.contextmanager
-def temp_filename(filename):
-    tmpdir = tempfile.mkdtemp()
+def catch_logs(logger_name, level=logging.WARNING):
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Unrelated comment - i've seen quite a few `catch_logs` implementations. I see one in `traits` as well. Maybe update the one in `traits` and make it standard across projects?

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 8]
```diff
@@ -15,59 +19,106 @@
...
+        logger.propagate = old_propagate
+        logger.removeHandler(handler)
 
 
 @contextlib.contextmanager
-def redirect_stdout_to(filename):
-    """ Redirect stdout to a file, in a way that's immune to changes
-    to Python's sys.stdout.
+def capture_stream(stream, callback=None):
+    """
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: same comment as `catch_logs` here

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 9]
```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
...
+        return [
+            obj for obj in gc.get_objects()
+            if isinstance(obj, type)
+        ]
+
+    def create_and_destroy_kernel(self):
+        kernel = InternalIPKernel()
+        kernel.init_ipkernel(gui_backend=None)
+        kernel.new_qt_console()
+        kernel.new_qt_console()
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: why do we need two calls to `new_qt_console` when creating a kernel?

Comment (rahulporuri) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Is this just a check to ensure that we *can* open multiple qt consoles when the kernel is running?

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 10]
```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
...
+    def test_ctrl_c_message_suppressed(self):
+        captured = []
+        with capture_stream(sys.__stdout__, callback=captured.append):
+            kernel = InternalIPKernel()
+            kernel.init_ipkernel(gui_backend=None)
+            kernel.shutdown()
+
+        self.assertEqual(len(captured), 1)
+        captured_stdout = captured[0].decode("utf-8")
+        self.assertNotIn("Ctrl-C will not work", captured_stdout)
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Just so I understand correcrltly, you're checking if a string doesn't exist within another string here? 

And can you comment on what `captured_stdout` is and maybe add a test for that?

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 5 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: It was the wrong fix. The original problem that this code was fixing was triggered by this line: 

https://github.com/enthought/envisage/blob/0950c2dfef6a5b44c24fff1b20d71b087198fb97/envisage/plugins/ipython_kernel/internal_ipkernel.py#L127

But that line should not have been there in the first place: the `exit_now` attribute is there to tell the shell to shut down under a running event loop, but by the time we reach this code the event loop has already been shut down. So the right fix was to remove this line (and make sure that any other necessary shell cleanup has taken place, which is what some of the rest of this PR does).

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 6 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: The `cleanup_connection_file` and `iopub_thread.stop` are still happening: see the `kernelapp` code. The code's just been moved to a more appropriate place (it's the `kernelapp` that sets this up, so it's the kernelapp's job to shut them down; we couldn't do that before because we were trying to work with the upstream kernelapp).

The `exit_now` code should never have been there, as discussed above.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 8 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: And the same reply. :-)

`apptools` may possibly be the right place for this sort of thing.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 9 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Yep. Just making sure that nothing goes wrong even when our kernel has multiple consoles attached. That's presumably a harder case than just a plain kernel with no consoles.

**Comment by mdickinson** [comment]
> There's still a lingering puzzle about three `IOPubThread` sockets: I can't figure out how/where they're being closed

FWIW, I tracked one of these (`IOPubThread.pipe_in`) down to this line: 

https://github.com/ipython/ipykernel/blob/18f2ef77b6a72109a1e50d8229e7216f1cfc2e39/ipykernel/iostream.py#L79

The zmq socket in question is wrapped in a zmq stream that's registered with the tornado io_loop, and the `self.io_loop.close(all_fds=True)` closes all the sockets that have been registered.

I didn't get as far as tracking down the other two, but I suspect they're similar; I'm content that the sockets are being closed properly.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 10 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Sorry, to be clear, yes, `captured_stdout` is a (Unicode) string, representing everything that was written out by the kernel. And yes, the `assertNotIn` is doing a string-in-string test to check that the Ctrl-C message suppression is working.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 2 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Done. And really, one `gc.collect()` should be enough (there are _theoretically_ situations where you need multiple `gc.collect()` calls to clean everything up, but this is unlikely to be one of those situations); let's avoid pandering to superstition and not use multiple calls until it becomes clear that that's necessary.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 4 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Done.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 3 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Done.

**Comment by rahulporuri** [comment]
I have no additional comments on this PR and any questions I had have been addressed.

File: envisage/plugins/ipython_kernel/heartbeat.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
+"""
+A simple ping-pong style heartbeat that runs in a thread.
+
+Modified from upstream to enable the heartbeat thread to be shut down cleanly.
+"""
+# We're currently using ipykernel v4.10.1. In that version, there's no way to
```
Comment (corranwebster) [inline envisage/plugins/ipython_kernel/heartbeat.py]: I don't think this requirement encoded anywhere other than implicitly what we get from EDM? There should be an optional setup.py requirement added for ipython support for folks getting things from `pip`.

File: envisage/plugins/ipython_kernel/heartbeat.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+class Heartbeat(ipykernel.heartbeat.Heartbeat):
+    """
+    A simple ping-pong style heartbeat that runs in a thread.
+
+    Modified from upstream to enable the thread to be shut down cleanly.
+    """
+    def run(self):
+        try:
+            super(Heartbeat, self).run()
+        except zmq.ZMQError as e:
```
Comment (corranwebster) [inline envisage/plugins/ipython_kernel/heartbeat.py]: Are other errors possible?  If so do we still want to clean up the socket before re-raising?

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 13]
```diff
@@ -3,17 +3,43 @@
...
+
+def log_print_err(*args, **kw):
+    """
+    Logging replacement for IPython.utils.io.raw_print_err.
+    Instead of printing to __stdout__, this function logs at WARNING level.
+    """
+    message = six.StringIO()
+    print(*args, sep=kw.get('sep', ' '), end=kw.get('end', '\n'),
+          file=message)
+    logger.warning(message.getvalue())
```
Comment (corranwebster) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: `warning` or `error`?

Also this method and `log_print` are awfully similar.

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py [hunk 14]
```diff
@@ -15,59 +19,106 @@
...
 
     from envisage.plugins.ipython_kernel.internal_ipkernel import (
         InternalIPKernel)
 
 
+# Machinery for catching logged messages in tests (both for testing
+# the existence and content of those messages, and for hiding the
+# logger output during the test run).
+
+class ListHandler(logging.Handler):
```
Comment (corranwebster) [inline envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: This feels generally useful, and may belong in `apptools.logging`.

File: envisage/plugins/ipython_kernel/heartbeat.py (hunk 11 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/heartbeat.py]: Ah, good point. I need to test whether this code actually works with all the latest versions from PyPI. It might, or it might not. If it does, great. If not, but there are easy fixes, I'll make those fixes. If not, and there are no easy fixes, we'll have to open issues and add version constraints in setup.py.

This *particular* hack should be forward-compatible: it'll work just as well with ipykernel 5.x, because that version already catches and deals with the exception, so our `except` block here won't get exercised.

Hmm; that does mean that the socket won't get deleted, though. I may still have to deal with that. Will check.

File: envisage/plugins/ipython_kernel/heartbeat.py (hunk 12 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/heartbeat.py]: Yes, we probably do; that should at least prevent the context.term call hanging. But we're in unknown territory at that point, and other bad things will probably happen.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 13 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Warning, I think, but I'm not 100% convinced either way. The implicit message here is "the IPython machinery spewed some stuff to stderr; here it is. It *might* be harmless, but it's probably worth investigating." That says "warning" rather than "error" to me.

It's probably academic anyway: I can't find a single use of `raw_print_err` (or its `rprinte` alias) within any of the IPython packages we're currently using.

Yes, they're similar, but they're so short that I think factoring out the repetition would make this harder to read rather than easier.

Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: I'm also considering using a custom logger name to make it clearer where the output is coming from in the logs. Something like `"ipython.stream_output"` instead of the current `"envisage.plugins.ipython_kernel.internal_ipkernel`, which is a bit obscure.

Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: One more thing: it may be worth adding a couple of direct tests for these two functions with bytes and Unicode arguments (and with a mixture of the two). It's all fine and clear cut on Python 3; less so on Python 2. If you replace `six.StringIO` with `io.StringIO`, for example, things will fail on Python 2, so there should be a test that also fails in that case.

**Comment by mdickinson** [comment]
> We'll need to figure out how to work around it here, but that can happen in another PR.

I think the easiest solution is to *not* towork around it and wait for ipykernel 5.1.3 (or later) to be released.

Merging!

File: envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py (hunk 14 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py]: Opened enthought/apptools#96
